### DISCUSSION TITLE
From DynamicImage to Vec<u8>

### DISCUSSION THREAD

**Discussion by nricciardi** [image-rs/image#2131] [Q&A]
I would use this image library to perform some elaboration over images. In particular, I would open an image, compress and save as base64. I would save it as base64 to include that in html page.
At the moment, I'm reading image using `File`:

```rust
let mut image = File::open(image_path.clone()).unwrap();

let mut buffer: Vec<u8> = Vec::new();
image.read_to_end(&mut buffer).unwrap();
```

and then I use `base64` library.

Now, I would use `image::load_from_memory` or `image::open` to obtain DynamicImage and then get the Vec<u8> from it.

I have tried with `as_flat_samples_u8().unwrap().samples` and `into_bytes()`, but the Vec of `image.read_to_end(&mut buffer)` is different from the Vec of `into_bytes()`. This last is longer.

How can I fix?

**Comment by fintelia** [comment]
You are probably looking for the `write_to` method:
```rust
let mut bytes: Vec<u8> = Vec::new();
img2.write_to(&mut Cursor::new(&mut bytes), image::ImageOutputFormat::Png)?;
```

If you call `into_bytes`, you get the raw pixel values of the image. While `write_to` encodes the image to a specific format (in this case PNG).

**Reply by nricciardi** [reply]
YES, this work!
But I don't know how... checking the two Vec<u8>, they are different, but after the encoding in base 64 and put the string in HTML, it works (I can see the image).

**Reply by fintelia** [reply]
Image encoding isn't deterministic. There are many different PNG's that all decode to the same thing. It generally comes down to a tradeoff between compression ratio and encoding time

**Reply by nricciardi** [reply]
Can DynamicImage encoded in different format without dump it in a file (I have seen `save_with_format`)?
I would compress my image in AVIF or WebP format and then create base 64 string of it.

**Reply by fintelia** [reply]
The `write_to` method doesn't dump it to a file. It just encodes the data into an image file format. That's a necessary step because web browsers expect the base64 string to be encoded in some such format.

If you want to switch to WebP or AVIF then you can change the `image::ImageOutputFormat` part of the sample accordingly. Though you might have to enable an additional crate feature

**Reply by nricciardi** [reply]
Yes, but it store "new image" in a Vec<u8>. I would replace my DynamicImage variable.

**Reply by fintelia** [reply]
I'm not sure what you're asking. If you want to get the backing `Vec` out of a `DynamicImage` then you can match on its variants and call `into_raw` on them. But it'll contain the raw pixel values rather than having them encoded in some image format.

Conversely, if you want the original bytes from the file, there's nothing wrong with your `File::open`/`read_to_end` approach.

**Reply by nricciardi** [reply]
I would open an image file and save it in a `DynamicImage` variable, then compress in WebP or AVIF to reduce base64 string length. So, I would override store new encoded image in `DynamicImage` type. Some like this:

```rust
let image: DynamicImage = image::load_from_memory(vec_u8_buffer);

image = image.into_avif();
// or .into_webp()
```

**Reply by fintelia** [reply]
A `DynamicImage` holds raw uncompressed pixel data. You cannot put AVIF or WebP data into it.
